Adds the get_or_create_patient function that looks for patients by MRN

fredkingham opened the pull request:
It looks by hosptial number locally, then by merged MRN locally.

If its not found it checks upstream whether this hospital number has been marked as in active and merged. If it has it returns the active MRN that it is merged with.

Commit:
Adds the get_or_create_patient function that looks for patients by MRN

It looks by hosptial number locally, then by merged MRN locally.

If its not found it checks upstream whether this hospital number has
been marked as in active and merged. If it has it returns the active MRN
that it is merged with.

Commit:
Adds in the get_patient which takes an MRN and returns the patient

If there is a patient with that MRN return that, otherwise see if there
is a patient that has that MRN as an inactive previously merged MRN and
return that.

Commit:
Changes the order of MRNs returned by get_mrn_and_date_from_merge_comment

We now return them with the most recent first

Commit:
Changes create_rfh_patient_from_hospital_number to created MergedMRN

create_rfh_patient_from_hospital_number now looks through the upstream database for the assosicated MRNs.
Creates an MRN for whichever is the active MRN.
Creates MergedMRNs for any inactive MRNs.

Returns the patient.

It also changes utils.get_or_create patient to not check upstream merged
mrns as this is no longer necessary.

Commit:
Merge branch 'cerner-merge-branch' into merge-get-or-create-patient

davidmiller commented on intrahospital_api/loader.py:
```diff
@@ -99,16 +101,54 @@ def create_rfh_patient_from_hospital_number(hospital_number, episode_category, r
     )
+    merged_mrns = []
+    for row in rows:
+        if row["ACTIVE_INACTIVE"] == emodels.MergedMRN.INACTIVE:
```
Does this work ? I thought this would give you the active MRN not the merged one ?

davidmiller commented on intrahospital_api/loader.py:
```diff
@@ -99,16 +101,54 @@ def create_rfh_patient_from_hospital_number(hospital_number, episode_category, r
+                upstream_merge_datetime=merge_dt,
+                merge_comments=row["MERGE_COMMENTS"]
+            ))
+    emodels.MergedMRN.objects.bulk_create(merged_mrns)
```
"There are never more than 3" <- this would be an example of a time I think it's not worth the container datastructure to bulk create

davidmiller commented on elcid/utils.py:
```diff
@@ -81,11 +82,30 @@ def natural_keys(text):
     return [ atoi(c) for c in re.split(r'(\d+)', text) ]
 
 
+def get_or_create_patient(mrn, episode_category, run_async=None):
```
Where do we think this is going to be called?

davidmiller commented on elcid/utils.py:
```diff
@@ -81,11 +82,30 @@ def natural_keys(text):
 
 
+def get_or_create_patient(mrn, episode_category, run_async=None):
+    from intrahospital_api import loader
```
Honestly I'd be tempted to put this in the intrahospital_api if we're doing this kind of thing

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -194,6 +195,69 @@ def update_if_changed(instance, update_dict):
+    return sorted(result, key=lambda x: x[1], reverse=True)
+
+
+def get_related_rows_for_mrn(mrn):
```
Is this really what you want? Is there a reason I need more than `fn(mrn)` ... `return primary, [merged_mrn, merged_mrn,...]` where primary and merged_mrn are character strings with the MRN value?

Commit:
Moves the get_or_create_patient function to the intrahospital_api.loader

Commit:
Removes the bulk_create for merged mrns in get_or_create_patient

Commit:
Changes the api for extracting merge data from the CRS_Patient_Masterfile

Previously we had update_demographics.get_related_rows_for_mrn which
returned all rows in the CRS_Patient_Masterfile related to an MRN.

Now we have get_active_mrn_and_merged_mrn_data which returns the active
MRN and then the data to created MergedMRN objects.

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -194,6 +195,103 @@ def update_if_changed(instance, update_dict):
+    If there are no related rows, ie the patient
+    is not merged, return None.
+    """
+    query = """
```
Could be convinced otherwise, but I _think_ I quite like the convention of queries all together at the top of the file in loaders as global variables. I think it's running debugging queries where it's particularly helpful

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -194,6 +195,103 @@ def update_if_changed(instance, update_dict):
+        related_related_mrns = [
+            i[0] for i in get_mrn_and_date_from_merge_comment(related_result["MERGE_COMMENTS"])
+        ]
+        for related_related_mrn in related_related_mrns:
```
Use recursion ? 
(If nesting can happen one level, is there any reason it can't happen a second level?)

davidmiller commented on intrahospital_api/update_demographics.py:
```diff
@@ -194,6 +195,103 @@ def update_if_changed(instance, update_dict):
+    # if there is not exactly 1 active row, we don't know what to merge
+    # so do nothing.
+    if not len(active_rows) == 1:
+        return mrn, []
```
I'm not sure 'do nothing' is exactly the right thing to do - are we aware how often this occurs?

fredkingham commented:
So the MergedMRN model looks like: 

class MergedMRN(models.Model):
    patient = models.ForeignKey(Patient, on_delete=models.CASCADE)
    mrn = models.CharField(max_length=256)
    merge_comments = models.TextField(blank=True, null=True, default="")
    upstream_merge_datetime = models.DateTimeField(blank=True, null=True)
    our_merge_datetime = models.DateTimeField(blank=True, null=True)

So in the simple case of e.g. MRN 123 has the merge comment "Merged with MRN 345 Oct 17 2016 11:03AM" where 123 is inactive and 345 is active this works. We get 
```py
MergedMRN(
   patient=Patient(345), 
   mrn=123, 
   merge_comment="Merged with MRN 345 Oct 17 2016 11:03AM", 
   upstream_merge_datetime=17/10/2016
  )
```
complex case:
 123, inactive, "Merged with 234 on Oct 17 2016 11:03AM"
 234, inactive, "Merged with 123 on Oct 17 2016 11:03AM, merged with 345 on Oct 18 2018 11:03AM"
 345, active, "Merged with 234 on Oct 18 2018 11:03AM"
 
 We get
 ```py
[
    MergedMRN(
        patient=Patient(345), 
        mrn=123,
        merge_comment="Merged with 234 on Oct 17 2016 11:03AM", 
        upstream_merge_datetime=17/10/2016
   ),
    MergedMRN(
        patient=Patient(345), 
        mrn=234,
        merge_comment="Merged with 234 on Oct 17 2016 11:03AM, merged with 345 on Oct 18 2018 11:03AM" 
        upstream_merge_datetime=17/10/2016
    ),
    MergedMRN(
        patient=Patient(345), 
        mrn=234,
        merge_comment="Merged with 234 on Oct 17 2016 11:03AM, merged with 345 on Oct 18 2018 11:03AM", 
        upstream_merge_datetime=18/10/2016
    ),
  ]
```

The second and third rows are mapping the relationships from mrn 234 to merge comment to 123 and 345 respectively but are a bit useless as its not clear the MRN they are merging. The issue is the upstream datetime is a function of the deconstructed merge comment, while the rest is that mapping of an MRN to an active MRN

Commit:
Recursively search through the merge comments.

This change changes it so that we recursively iterate through merge
results.

i.e. if MRN 123 is merged into MRN 234 which is merged into 345 which is
merged into 456, we created a MergedMRN object for 123 -> 456.

Prior to this we only went down one level.

This change also changes the definition of a MergedMRN. Prior to this is
what had a quasi relationship where it was inbetween a merge comment and
a row in the master file.

Now it has a 1-1 relationship with a row in the master file.

fredkingham commented:
I have decided to change the MergedMRN to be explicitly defined as a row in the Cerner masterfile for an MRN that is inactive. 

This means that the upstream_merge_datetime has a fuzzy meaning as there can be 2 merges for a single MRN into inactive rows. The upstream_merge_datetime is only used for ordering the merged MRNs in the search results so approximate is ok. I've changed the upstream_merge_datetime to be the highest datetime mentioned in the merge comment.

Needs discussion as If you're happy with this approach it would be more sensible to delete upstream_merge_datetime and move that logic into the search results

Commit:
Removes the ACTIVE/INACTIVE constants from the MergedMRN model.

These are actually just features of the masterfile and are not connected
to the model specifically. Indeed they are not used by any field on the
model.

I've changed the ACTIVE flag to be hardcoded when we load in from the
masterfile and removed the INACTIVE flag as it is never used.

Commit:
Backs out the changes to elcid/utils.py related to get_or_create_patient

We used to have get_or_create_patient in elcid.utils. This was removed
from a previous change however some of the imports remained. This
backs out the changes to imports.

Commit:
Moves the query used by get_masterfile_row to the top of the file.

Our convention is that queries are held in constant case strings at the
top of the file. This changes
intrahospital_api.update_demographics.get_masterfile_row to use this
convention

Commit:
Changes get_masterfile_row into get_merged_masterfile_row

intrahospital_api.update_demographics.get_masterfile_row only returned
rows that had been merged. This changes the function name to reflect
that.

Commit:
cleans up the doc string for crawl_merge_comments

Commit:
Improves a comment in crawl_merge_comments

Commit:
Changes get_active_mrn_and_merged_mrn_data to handle and MRN with no merges

If there are no merges get_active_mrn_and_merged_mrn_data returns the
current MRN as the active MRN and an empty list for the MRNs that have
been merged into it.

Commit:
Fixes error handling of get_active_mrn_and_merged_mrn_data

We should not fail to load an MRN if there are discrepencies with the
upstream data however we should get an email (sent by logger.error)

Commit:
Updates the doc string for get_active_mrn_and_merged_mrn_data

Updates the doc string to include how we handle internal issues in the
merge statement.

fredkingham commented:
complex case:
123, inactive, "Merged with 234 on Oct 17 2016 11:03AM"
234, inactive, "Merged with 123 on Oct 17 2016 11:03AM, merged with 345 on Oct 18 2018 11:03AM, merged with 456 Oct 19 11:03 AM,"
345, inactive, "Merged with 234 on Oct 18 2018 11:03AM, Merged with 456 20 Oct 2018 11:03AM"
456, active, "Merged with 234 Oct 19 11:03 AM, Merged with 345, 20 Oct 2018 11:03AM"

Commit:
Removes all references to upstream_merge_datetime from update_demographics

We have removed the field upstream_merge_datetime from MergedMRN on commit cde1e75e9900406ab69f4b3a69bfd7c542e45349

fredkingham commented:
As discussed, we've removed upstream_merge_datetime

Commit:
Makes the doc string on the MergedMRN clearer

Commit:
Renames crawl_merge_comments to recursively_parse_merge_comments

Commit:
mocks load_patint in CreateRfhPatientFromHospitalNumberTestCase

This means we are checking that load_patient is not called when an
exception is raised.

It also removes a large amount of log statments when running tests.

davidmiller committed:
Fixup: Docstring for MergedMRN

davidmiller committed:
Fixup: .exists() is much clearer than implicit queryset truthiness

Commit:
Merge branch 'cerner-merge-branch' into merge-get-or-create-patient

Commit:
Merge branch 'merge-get-or-create-patient' of github.com:openhealthcare/elcid-rfh into merge-get-or-create-patient

Commit:
Changes get_active_mrn_and_merged_mrn_data to deal with exceptions

There are 3 sorts of issue we can have with
get_active_mrn_and_merged_mrn_data.

1. We get multiple rows for a single MRN in the masterfile.
This should never happen. In this instance we raise a ValueError because we
should not proceed in this situation.

2. The patient does not exist in the CernerMasterFile.
This should usually not happen. There are cases where we have patients in our
system that are not registered  in the RFH Cerner, these situations need
to be handled by the calling code.

3. There is a typo in the structure of the merge information.
We have instances where there are multiple active patients and instances
where there are only inactive instances. In this case we should allow
the creation of the patient and send us an email so that we can look
into it.

Commit:
Fixes flake8 formatting issues in the intrahospital_api.loader

Commit:
Adds in the boolean rfh_patient argument to get_or_create_patient

Sometimes we deal with patients who are not RFH patients for example
patients from other hospitals who our doctors are advising on.

In these cases they should be able to create a patient but we should not
look for RFH data such as imaging for them.

This argument tells the calling code to not look in upstream databases
for information about the patient.

fredkingham commented:
Todays changes do 2 things.

1. We raise different exceptions on different problems. 
     * MergeException if we should allow the patient to be created but that we should be notified
     * CernerPatientNotFoundException if the patient is not found in the master file. We should not create the patient and raise an error. This can be handled by the calling code.
     * ValueError if there are multiple rows for an MRN in the MasterFile. This should never be the case and to are knowledge has never been the case. We should probably stop the process if this has happened as it means something really wrong has happened to the upstream data.
     
2. `get_or_create_patient` now accepts an rfh_patient argument (default=True). 
We discussed RNOH patients and how they would not exist in the CernerMasterFile. I am assuming they also would not have lab/imaging data? If we pass rfh_patient=False then we allow a patient to be created without checking demographics to make sure they exist and without the loaders running.

This leads us to the situation where create_rfh_patient_from_hospital_number takes an argument called `rfh_patient` patient which is obviously not great but we are planning to rename this function in a future PR where we will add __ and remove the reference to rfh in the name.

davidmiller committed:
There's nothing merged about either this query or function

davidmiller committed:
While implementation of parse_merge_comments

davidmiller committed:
Functions claiming to return things should include a return statement

davidmiller committed:
Fixup pro tip: Don't access variables that are created on the following line.

davidmiller committed:
Fixup: append not assign

Commit:
Fixes update_demographics.parse_merge_comments to assign the correct active MRN

It was assigning the MRN that was passed in as the active MRN. This
commit changes it so that it assigns the next_mrn variable. It renames
the mrn variable to initial_mrn as we are dealing with a lot of MRNs so
hopefully this makes it easier to read.

Commit:
Fixes unit tests for update_demographics.parse_merge_comments

davidmiller committed:
Merge pull request #1414 from openhealthcare/recursion-adjustments2

Recursion adjustments2
